pkg/kube/krt/collection.go

```diff
@@ -473,11 +485,14 @@ func (h *manyCollection[I, O]) Register(f func(o Event[O])) Syncer {
 	return registerHandlerAsBatched[O](h, f)
 }
 
-func (h *manyCollection[I, O]) RegisterBatch(f func(o []Event[O]), runExistingState bool) Syncer {
-	if !h.eventHandlers.Insert(f) && runExistingState {
-		// Already started. Pause everything, and run through the handler.
+func (h *manyCollection[I, O]) RegisterBatch(f func(o []Event[O], initialSync bool), runExistingState bool) Syncer {
+	if runExistingState {
 		h.recomputeMu.Lock()
 		defer h.recomputeMu.Unlock()
+	}
+	init := h.eventHandlers.Insert(f)
```

Comment:
Isn't this always return false as `MarkInitialized` is called before `RegisterBatch`

In addition, the `init` =  !o.init, should be negative

Reply:
> Isn't this always return false as MarkInitialized is called before RegisterBatch



Its not. RegisterBatch can be called from other places. The specific race is for things _outside_ krt calling RegisterBatch and krt itself calling RegisterBatch, which was resulting in double events.

Good call on the naming, will fix